### ISSUE TITLE
Verify entropy from EDN is used correctly

### ISSUE THREAD

**Issue by weicaiyang** [lowRISC/opentitan#8458]
@tjaychen 

As we discussed, it's good to prove EDN data are used to wipe the secrets in keymgr.

> EDN -> LFSR -> used to wipe secret/data output

From EDN to the input of LFSR, we can check EDN data are sent to LSFR as seeds. But after LFSR, the value is scrambled. Sometimes, it's constantly scrambled for a period of time. Hence it's hard for DV to predict the correct outcome.

**Comment by tjaychen** [comment]
o you mean inside prim_lfsr the scrambling that's done?

**Comment by weicaiyang** [comment]
> o you mean inside prim_lfsr the scrambling that's done?

yeah, need to know how the EDN input is scrambled by prim_lfsr and then used to wipe data. In invalid state, SW output and sideload output are wiped by random data from LFSR. If I can predict these random data based on the EDN input, then I can do end to end comparison.

**Comment by tjaychen** [comment]
@weicaiyang i feel like this one is either nice to have / backlog at this point... i feel like trying to model how the lfsr scrambles in dv is probably not worth it at this point... let me know what you think.

**Comment by weicaiyang** [comment]
> @weicaiyang i feel like this one is either nice to have / backlog at this point... i feel like trying to model how the lfsr scrambles in dv is probably not worth it at this point... let me know what you think.

agree, I still can't find a good way to verify this. Changed to backlog.

**Comment by vogelpi** [comment]
Hey @ballifatih , I think that based on the description this PR actually matters for keymgr. I've now added the corresponding label and updated the title.

**Comment by ballifatih** [comment]
Thanks @vogelpi, good catch!

Triaged for `keymgr`. I think we should include this in M2.5.

It might be a bit challenging to get the exact LFSR state that is written onto the cleared key, because when the clear mode is activated, it continuously wipes the key (see [the thread](https://github.com/lowRISC/opentitan/pull/16492#issuecomment-1331504542)). Because we toggle the clearing on and off from SW side, I do not think that the number of clock cycles is fixed. 

If we cannot do this ideal test, we should probably at least that the initial seed that is loaded into keymgr's LFSR is correct.

**Comment by vogelpi** [comment]
Yes, I agree with you @ballifatih . It's probably not worth to predict the correct LFSR state/output. What we could do the following with SVAs in the design:
1. Verify that EDN input is indeed loaded into LFSR state.
2. Verify that after wiping, the key register indeed changes to a different value / the LFSR output.
Both this is already done for AES, too. I can give you a pointer if needed. Just let me know when you start working on this.

**Comment by ballifatih** [comment]
When I try to replicate [the SVAs from AES](https://github.com/lowRISC/opentitan/blob/master/hw/ip/aes/dv/sva/aes_masking_reseed_if.sv), I see two irregularities for keymgr:

1. Unlike AES, where there are multiple independent 32-bit LFSR chunks, keymgr consist of single 64-bit LFSR. Two 32-bit reads are combined with `prim_packer_fifo` inside `prim_edn_req` and then 64-bit is written to LFSR. Note that, `LfsrWidth = 64`:
https://github.com/lowRISC/opentitan/blob/060354c410761db905f385909c2b163b0076258d/hw/ip/keymgr/rtl/keymgr_reseed_ctrl.sv#L60-L75
2. When there is an ongoing operation in keymgr, LFSR update operation is ignored. This means if we write an SVA with the assumption that every valid `prim_edn_req` output is written to LFSR, we would see some violations.
https://github.com/lowRISC/opentitan/blob/060354c410761db905f385909c2b163b0076258d/hw/ip/keymgr/rtl/keymgr.sv#L215-L220

What I can think of is to provide partial coverage with SVAs that capture:

1. Ensure that what comes out of `prim_edn_req` (64-bit) is written to LFSR.
2. Ensure that every even 32-bit read from EDN comes out as one half of `prim_edn_req`
3. When there is key wipe, ensure that key is updated with something new (I am guessing it would be difficult to guess which state LFSR evolves to at that point)

WDYT @vogelpi ?

**Comment by ballifatih** [comment]
> estimate 32

**Comment by vogelpi** [comment]
@vogelpi to take a look again.

**Comment by vogelpi** [comment]
I think we should try to include this for M5. The strategy proposed by @ballifatih above sounds good to me (sorry for not getting back to you Fatih, I completely forgot about this one obviously).

Would this be something you could help with @ballifatih for M5?

**Comment by moidx** [comment]
For M5 sign-off we are planning to follow an RTL review by inspecting the waveforms. 

@ballifatih will take a look at the waveforms. 

@vogelpi, someone else should help with implementing the assertions. It is probably ok to move the assertions to M7 and rely on the RTL/waveform auditing for M5.

**Comment by vogelpi** [comment]
Thanks for the doing the inspection and reporting back the results @ballifatih , this is perfect!

I am now moving this to M7 for adding the SVAs as proposed by @moidx . I've increased the effort spent by 0.5 to record your work Faith.
